Apps/Clouseau/DESIGN.org

```diff
@@ -0,0 +1,258 @@
+     loop.
+
+  3. Places in which objects reside should be modeled explicitly and,
+     to some extend, indicated to the user. The distinction between an
+     object and the place it resides in can be important for a number
+     of reasons:
+
+     + Changes and immutable objects: incrementing a variable bound to
+       a number is a common operation which would not be expressible
+       in a model consisting only of values since numbers are
+       immutable. In contrast, incrementing a place such as a slot in
+       an instance or a cell in an array is fine, even if the value of
+       the place is an immutable object.
+
+     + Some places can be unbound and can be made unbound. A model
+       consisting only of values has no way of expressing properties
+       of the place in which a value resides.
+
+     The previous inspector architecture generally only represented
+     objects, making it impossible for a command applied to an object
+     to, for example, retrieve an array containing the object and
+     modify the array.
+
+  4. Each inspected object should have an associated state that
+     persists across redisplays and can store information specific to
+     the object class (E.g. instance slots can be displayed as a flat
+     list or organized by superclass from which the slot has been
+     inherited). This mechanism should be extensible.
+
+     The previous inspector architecture associated pre-defined pieces
+     of state such as the expansion status or style in which cons
+     trees should be displayed with objects. This mechanism was not
+     extensible.
+
+  5. It should be possible to define commands which are only
+     applicable to specific object classes.
+
+     The previous inspector architecture did not generally use
+     specific presentation types and therefore could not control the
+     applicability of commands very well.
+
+* Design
+
+** Concepts
+
+   According to the above [[*Requirements][requirements]], the following concepts are
+   central to the new architecture:
+
+   + place :: Semantically similar to the Common Lisp term, but
+              first-class. A place can have a value or may be
+              unbound. Some kinds of places allow changing the value
+              and/or making the place unbound.
+
+   + object state :: Information associated with an inspected object and
+                     the place the object resides in. If the same
+                     (~eql~) object resides in multiple places, each
+                     place-object pair has its own object state.
+
+                     Object states are ~present~ ed with presentation
+                     type ~inspected-object~. Since there can be
+                     specific object state classes for place
+                     class-object class combinations, presentation
+                     methods can be specialized to object state
+                     classes.
```

Comment:
I think that this paragraph could be rephrased. If something is a class it is implicitly a presentation type (a reverse implication does not hold). I would focus on the class aspect of the object state and describe it as a protocol class which may be subclassed (and presentation methods may be specialized on subclasses).

Reply:
Fixed.